### PR TITLE
chore: simplify master replication cancelation interface

### PR DESCRIPTION
**Pull request by romange**
Before that CancelReplication did too many things, moreover, we had StopReplication that did the same.

This PR moves CancelReplication under ReplicaInfo struct, and reduces code duplication around this change.

### CODE REVIEW

File: src/server/dflycmd.cc [hunk 1]
```diff
@@ -99,6 +99,32 @@ bool WaitReplicaFlowToCatchup(absl::Time end_time, const DflyCmd::ReplicaInfo* r
 
 }  // namespace
 
+void DflyCmd::ReplicaInfo::Cancel() {
```
Comment (romange) [inline src/server/dflycmd.cc:102]: moving the first code block from CancelReplication.

**Review by dranikpg — APPROVED** [review]
👍🏻

File: src/server/dflycmd.cc [hunk 2]
```diff
@@ -99,6 +99,32 @@ bool WaitReplicaFlowToCatchup(absl::Time end_time, const DflyCmd::ReplicaInfo* r
...
+  lock_guard lk = GetExclusiveLock();
+  if (replica_state == SyncState::CANCELLED) {
+    return;
+  }
+
+  LOG(INFO) << "Disconnecting from replica " << address << ":" << listening_port;
+
+  // Update tate and cancel context.
```
Comment (chakaz) [inline src/server/dflycmd.cc:110]:
```suggestion
  // Update state and cancel context.
```

File: src/server/dflycmd.cc [hunk 3]
```diff
@@ -99,6 +99,32 @@ bool WaitReplicaFlowToCatchup(absl::Time end_time, const DflyCmd::ReplicaInfo* r
...
+  cntx.Cancel();
+
+  // Wait for tasks to finish.
+  shard_set->RunBlockingInParallel([this](EngineShard* shard) {
+    FlowInfo* flow = &flows[shard->shard_id()];
+    if (flow->cleanup) {
+      flow->cleanup();
+    }
+
+    flow->full_sync_fb.JoinIfNeeded();
```
Comment (chakaz) [inline src/server/dflycmd.cc:121]: Do we want to block the shard queue on that?

Comment (romange) [reply src/server/dflycmd.cc:121]: 1. it does not run in shard queue but in a dedicated fiber.
2. it's the same code as before, just refactored into a function

**Review by chakaz — APPROVED** [review]

**Pull request merged**